hare3/hare.go

```diff
@@ -0,0 +1,571 @@
+	ctx    context.Context
+	inputs chan<- *sessionInput
+}
+
+func (inputs *sessionInputs) submit(msg *input) (*response, error) {
+	input := &sessionInput{
+		input:  msg,
+		result: make(chan *response, 1),
+	}
+	select {
+	case <-inputs.ctx.Done():
+		return nil, inputs.ctx.Err()
+	case inputs.inputs <- input:
+	}
+	select {
+	case resp := <-input.result:
+		return resp, nil
+	case <-inputs.ctx.Done():
+		return nil, inputs.ctx.Err()
+	}
+}
+
+type Opt func(*Hare)
+
+func WithWallclock(clock clock.Clock) Opt {
+	return func(hr *Hare) {
+		hr.wallclock = clock
+	}
+}
+
+func WithConfig(cfg Config) Opt {
+	return func(hr *Hare) {
+		hr.config = cfg
+		hr.oracle.config = cfg
+	}
+}
+
+func WithLogger(logger *zap.Logger) Opt {
+	return func(hr *Hare) {
+		hr.log = logger
+		hr.oracle.log = logger
+	}
+}
+
+// WithEnableLayer can be used to pass genesis layer.
+// Note that for it to have effect it needs to be after WithConfig.
+func WithEnableLayer(layer types.LayerID) Opt {
+	return func(hr *Hare) {
+		hr.config.EnableAfter = layer
+		hr.oracle.config.EnableAfter = layer
+	}
+}
+
+func WithTracer(tracer Tracer) Opt {
+	return func(hr *Hare) {
+		hr.tracer = tracer
+	}
+}
+
+func New(
+	nodeclock *timesync.NodeClock,
+	pubsub pubsub.PublishSubsciber,
+	db *datastore.CachedDB,
+	verifier *signing.EdVerifier,
+	signer *signing.EdSigner,
+	oracle oracle,
+	sync system.SyncStateProvider,
+	opts ...Opt,
+) *Hare {
+	ctx, cancel := context.WithCancel(context.Background())
+	hr := &Hare{
+		ctx:      ctx,
+		cancel:   cancel,
+		results:  make(chan ConsensusOutput, 32),
+		coins:    make(chan WeakCoinOutput, 32),
+		sessions: map[types.LayerID]sessionInputs{},
+
+		config:    DefaultConfig(),
+		log:       zap.NewNop(),
+		wallclock: clock.New(),
+
+		nodeclock: nodeclock,
+		pubsub:    pubsub,
+		db:        db,
+		verifier:  verifier,
+		signer:    signer,
+		oracle: &legacyOracle{
+			log:    zap.NewNop(),
+			oracle: oracle,
+			config: DefaultConfig(),
+		},
+		sync:   sync,
+		tracer: noopTracer{},
+	}
+	for _, opt := range opts {
+		opt(hr)
+	}
+	return hr
+}
+
+type Hare struct {
+	// state
+	ctx      context.Context
+	cancel   context.CancelFunc
+	eg       errgroup.Group
+	results  chan ConsensusOutput
+	coins    chan WeakCoinOutput
+	mu       sync.Mutex
+	sessions map[types.LayerID]sessionInputs
+
+	// options
+	config    Config
+	log       *zap.Logger
+	wallclock clock.Clock
+
+	// dependencies
+	nodeclock *timesync.NodeClock
+	pubsub    pubsub.PublishSubsciber
+	db        *datastore.CachedDB
+	verifier  *signing.EdVerifier
+	signer    *signing.EdSigner
+	oracle    *legacyOracle
+	sync      system.SyncStateProvider
+	tracer    Tracer
+}
+
+func (h *Hare) Results() <-chan ConsensusOutput {
+	return h.results
+}
+
+func (h *Hare) Coins() <-chan WeakCoinOutput {
+	return h.coins
+}
+
+func (h *Hare) Start() {
+	h.log.Info("started", zap.Inline(&h.config))
+	h.eg.Go(func() error {
+		h.pubsub.Register(h.config.ProtocolName, h.Handler)
```

Comment:
i think there is a misunderstand. we are getting hare when node is not synced, they are simply dropped. in this code they will be dropped later, because session is not registered

Reply:
synced condition is applied in onLayer method